pkg/virt-handler/rest/console.go

```diff
@@ -0,0 +1,195 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package rest
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"path"
+	"strconv"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+type ConsoleHandler struct {
+	podIsolationDetector isolation.PodIsolationDetector
+	serialStopChans      map[types.UID](chan struct{})
+	vncStopChans         map[types.UID](chan struct{})
+	serialLock           *sync.Mutex
+	vncLock              *sync.Mutex
+	vmiInformer          cache.SharedIndexInformer
+}
+
+func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiInformer cache.SharedIndexInformer) *ConsoleHandler {
+	return &ConsoleHandler{
+		podIsolationDetector: podIsolationDetector,
+		serialStopChans:      make(map[types.UID](chan struct{})),
+		vncStopChans:         make(map[types.UID](chan struct{})),
+		serialLock:           &sync.Mutex{},
+		vncLock:              &sync.Mutex{},
+		vmiInformer:          vmiInformer,
+	}
+}
+
+func (t *ConsoleHandler) VNCHandler(request *restful.Request, response *restful.Response) {
+	vmi, err := t.getVMI(request)
+	if err != nil {
```

Comment:
If a VMI does not exist, return `404`. If the VMI can't be retrieved from the cache, return `500`.

Reply:
are you sure? I don't mind doing that, but I would go with `500` in both cases - after all, the vmi existed at the api-level at the time the user invoked this opeation..

Reply:
It is very unlikely but possible to hit the scenario where it was still ok and there when you hit the api, but disappeared afterwards. Or not?

Reply:
I think it may happen, all I'm saying is that in this case the error would be `500` because as a user I invoked the operation on a valid entity - the server later failed to execute the operation

Reply:
From virt-handler perspective it is 404 (not 500). It was also valid from virt-api to forward the connection, based on its information. Looks to me like virt-api should return 404?

Reply:
I think it is important that also for internal REST-communication it is important to send the right error codes. Sending 500 on an action which is perfectly valid will look like a bug.

Reply:
ok, if virt-api is not expected to forward this code to the client then I agree

Reply:
Done